[WebXR] Warn when accessing XRWebGLLayer framebuffer dimensions before first render

djg opened the pull request:
#### 227c6a7e9cad3099da4c496dbc311de75808f9b8
<pre>
[WebXR] Warn when accessing XRWebGLLayer framebuffer dimensions before first render
<a href="https://bugs.webkit.org/show_bug.cgi?id=299934">https://bugs.webkit.org/show_bug.cgi?id=299934</a>
<a href="https://rdar.apple.com/161708558">rdar://161708558</a>

Reviewed by Mike Wyrzykowski.

Add console warnings when developers access framebufferWidth or
framebufferHeight properties on XRWebGLLayer before the framebuffer has been
properly initialized.  This commonly occurs when accessing these properties
outside of a requestAnimationFrame callback, where the framebuffer dimensions
are not yet available.

The warning message clarifies that accurate dimensions are unavailable until
requestAnimationFrame processing and that the API returns 1 as a fallback value.
This helps developers understand the timing requirements for accessing XR
framebuffer dimensions.

* Source/WebCore/Modules/webxr/WebXRWebGLLayer.cpp:
(WebCore::WebXRWebGLLayer::framebufferWidth const): Add warning when framebuffer size is empty.
(WebCore::WebXRWebGLLayer::framebufferHeight const): Add warning when framebuffer size is empty.
(WebCore::WebXRWebGLLayer::addConsoleMessage const): New helper method to log console messages.
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.h: Add addConsoleMessage declaration.

Canonical link: <a href="https://commits.webkit.org/300864@main">https://commits.webkit.org/300864@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2258b1f31f452eed4ea18755304364f9fa14bd57

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124155 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43848 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34565 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130982 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/99398226-27a3-4b0d-85f3-96f92492eade) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126032 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44589 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52448 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94440 "") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/11f71e8f-57b0-4e9c-852f-66f72004649d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127109 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35528 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111054 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75031 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/83f41015-a39e-410a-b81b-502c29c3ad1e) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34473 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29216 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74468 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105269 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29437 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133656 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51079 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/38927 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102911 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51464 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107272 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102718 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26124 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48074 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26322 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47959 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50939 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56714 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/50387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/53740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/52064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/Modules/webxr/WebXRWebGLLayer.cpp:
```diff
@@ -196,8 +197,12 @@ const WebGLFramebuffer* WebXRWebGLLayer::framebuffer() const
+    if (m_framebuffer) {
+        auto framebufferSize = m_framebuffer->drawFramebufferSize();
+        if (framebufferSize.isEmpty())
+            addConsoleMessage(MessageLevel::Warning, "accurate framebufferWidth is unavailable until requestAnimationFrame processing; returning 1"_s);
```
Had no idea we had this functionality to log JavaScript console warnings 👍

djg committed:
[WebXR] Warn when accessing XRWebGLLayer framebuffer dimensions before first render
https://bugs.webkit.org/show_bug.cgi?id=299934
rdar://161708558

Reviewed by Mike Wyrzykowski.

Add console warnings when developers access framebufferWidth or
framebufferHeight properties on XRWebGLLayer before the framebuffer has been
properly initialized.  This commonly occurs when accessing these properties
outside of a requestAnimationFrame callback, where the framebuffer dimensions
are not yet available.

The warning message clarifies that accurate dimensions are unavailable until
requestAnimationFrame processing and that the API returns 1 as a fallback value.
This helps developers understand the timing requirements for accessing XR
framebuffer dimensions.

* Source/WebCore/Modules/webxr/WebXRWebGLLayer.cpp:
(WebCore::WebXRWebGLLayer::framebufferWidth const): Add warning when framebuffer size is empty.
(WebCore::WebXRWebGLLayer::framebufferHeight const): Add warning when framebuffer size is empty.
(WebCore::WebXRWebGLLayer::addConsoleMessage const): New helper method to log console messages.
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.h: Add addConsoleMessage declaration.

Canonical link: https://commits.webkit.org/300864@main
